content/articles/How-to-start-up-with-Socket-Programming-with-Java/index.md

```diff
@@ -0,0 +1,315 @@
+
+```java
+Scanner scanner = new Scanner(System.in);
+```
+`System.in` specifies that we are getting the keyboard input.
+
+#### Methods used in Client Socket
+- `public InputStream getInputStream()` - returns an InputStream object for receiving data. Our Server will keep on receiving messages until client sends "BYE".
+- `public OutputStream getOutputStream()`- returns an OutputStream for sending data.
+- `BufferReader()` - Wraps inputStreamReader to improve efficiency
+- `BufferWriter()` - Wraps outputStreamReader to improve efficiency
+- `BufferedWriter.flush()` - Flushes the streams. The flush method is invoked when the buffer is complete.
+- `void close()`- closes all streams and the socket.
+  
+#### Full Client-Side Application
+```java
+package com. company;  
+  
+import java.io.*;  
+import java.net.Socket;  
+import java.util.Scanner;  
+  
+public class client {  
+ public static void main(String[] args) {  
+  Socket socket = null;  
+  InputStreamReader inputStreamReader = null;  
+  OutputStreamWriter outputStreamWriter = null;  
+  BufferedReader bufferedReader = null;  
+  BufferedWriter bufferedWriter = null;  
+  
+try {  
+socket = new Socket("localhost", 5000);  
+inputStreamReader = new InputStreamReader(socket.getInputStream());  
+outputStreamWriter = new OutputStreamWriter(socket.getOutputStream());  
+bufferedReader = new BufferedReader(inputStreamReader);  
+bufferedWriter = new BufferedWriter(outputStreamWriter);  
+Scanner scanner = new Scanner(System.in);  
+while (true)
+ {  
+  String msgToSend = scanner.nextLine();  
+  bufferedWriter.write(msgToSend);  
+  bufferedWriter.newLine();  
+  bufferedWriter.flush();  
+  
+  System.out.println("Server: " + bufferedReader.readLine());  
+  
+if (msgToSend.equalsIgnoreCase("BYE"))  
+    break;  
+  }  
+} catch (IOException e) {  
+   e.printStackTrace();  
+} finally {  
+try {  
+if (socket != null)  
+ socket.close();  
+if (inputStreamReader != null)  
+  inputStreamReader.close();  
+if (outputStreamWriter != null)  
+ outputStreamWriter.close();  
+if (bufferedReader != null)  
+ bufferedReader.close();  
+if (bufferedWriter != null)  
+ bufferedWriter.close();  
+} catch (IOException e) {  
+  e.printStackTrace();  
+          }  
+        }  
+    }  
+}
+```
+A sample output of this code:
+```
+Hey there 
+Server:  Message Received
+I am doing this for the first time
+Server:  Message Received
+BYE
+Server:  Message Received
+```
+### How to Run Programs in a Variety of Applications
+#### Using Intellij or another integrated development environment:
+- compile the two programs.
+- **Run** the server program first, then the client application.
+- Type messages in the client window, which will be received and shown by the server window at the same time.
+- To exit, type **BYE**.
+
+#### If you're using Command Prompt/Terminal:
+- Launch the server program as the java server, depending on how you saved it.
+- Run the client program on a terminal.
```

Comment:
You might have to go into detail about how to run it in a terminal. It's not a straightforward process for a beginner like it is in an IDE